src/dmd/dtemplate.d

```diff
@@ -8278,56 +8278,104 @@ MATCH matchArg(TemplateParameter tp, Scope* sc, RootObject oarg, size_t i, Templ
 }
 
 
+alias TemplateInstanceArray = Array!(void*);
+
 /***********************************************
  * Collect and print statistics on template instantiations.
  */
 struct TemplateStats
 {
-    __gshared TemplateStats[const void*] stats;
+    __gshared TemplateStats[const void*] g_stats;
+
+    uint totalInstanceCount = 0; // number of instantiations of the template
+    uint uniqueInstanceCount = 0; // number of unique instantiations of the template
 
-    uint numInstantiations;     // number of instantiations of the template
-    uint uniqueInstantiations;  // number of unique instantiations of the template
+    TemplateInstanceArray* uniqueInstances = null;
 
     /*******************************
      * Add this instance
      */
-    static void incInstance(const TemplateDeclaration td)
+    static void incInstance(const TemplateDeclaration td,
+                            const TemplateInstance ti)
     {
-        if (!global.params.vtemplates)
-            return;
-        if (!td)
-            return;
-        if (auto ts = cast(const void*) td in stats)
-            ++ts.numInstantiations;
+        if (!global.params.vtemplates) return;
+        assert(td);
+        assert(ti);
+        if (auto ts = cast(const void*) td in g_stats)
+        {
+            ++ts.totalInstanceCount;
+        }
         else
-            stats[cast(const void*) td] = TemplateStats(1, 0);
+        {
+            g_stats[cast(const void*) td] = TemplateStats(1, 0);
+        }
     }
 
     /*******************************
      * Add this unique instance
      */
-    static void incUnique(const TemplateDeclaration td)
+    static void incUnique(const TemplateDeclaration td,
+                          const TemplateInstance ti)
     {
-        if (!global.params.vtemplates)
-            return;
-        if (!td)
-            return;
-        if (auto ts = cast(const void*) td in stats)
-            ++ts.uniqueInstantiations;
+        if (!global.params.vtemplates) return;
+        assert(td);
+        assert(ti);
+        if (auto ts = cast(const void*) td in g_stats)
+        {
+            if (ts.uniqueInstances is null)
+            {
+                ts.uniqueInstances = new TemplateInstanceArray();
+            }
+            ts.uniqueInstances.push(cast(void*)ti);
+            ++ts.uniqueInstanceCount;
+        }
         else
-            stats[cast(const void*) td] = TemplateStats(0, 1);
+        {
+            g_stats[cast(const void*) td] = TemplateStats(0, 1);
+        }
     }
 }
 
+struct TdStats
+{
+    void *td;
+    TemplateStats ts;
+}
 
 void printTemplateStats()
 {
     if (!global.params.vtemplates)
         return;
 
-    printf("  Number   Unique   Name\n");
-    foreach (td, ref ts; TemplateStats.stats)
+    // printf("  Number   Unique   Name\n");
+    Array!(TdStats) sortedStats;
+    foreach (td_, ref ts; TemplateStats.g_stats)
+    {
+        sortedStats.push(TdStats(cast(void*)td_, ts));
+    }
```

Comment:
Should reserve the expected size beforehand

Reply:
Fixed.